src/corona_hakab_model/affinity_matrix.py

```diff
@@ -212,17 +88,169 @@ def normalize(self):
+        amount_of_connections = map(math.ceil, np.random.exponential(scale_factor - 0.5, amount_of_agents))
+
+        # dict of agents to amount of remaining connections to make for this agent
+        remaining_contacts: Dict[Agent, int] = {agent: amount for (agent, amount) in
+                                                zip(agents_to_use, amount_of_connections)}
+
+        # will be used as a stopping sign
+        connections_sum = sum(remaining_contacts.values())
+
+        # a list of indexes of agents which still lack connections. used for efficiency
+        available_agents = [agent for agent in agents_to_use]
+        available_agents_len = len(available_agents)
+
+        # pre-rolling all rolls for efficiency. for each connection, the 2nd agent will be rolled using this
+        # todo make sure the later-used % operator doesn't harm the randomness
+        rolls = np.random.randint(0, amount_of_agents, connections_sum // 2 + 1)
+        roll_cnt = 0
+
+        # this structure will be used to save all connections, and later on insert them all at once. used for efficiency
+        connections_tuples = np.zeros((2, connections_sum), dtype=np.int)
+        connections_cnt = 0
+
+        # while there are still connections left to make
+        while available_agents_len >= 2 and connections_sum >= 2:
+
+            current_agent = available_agents.pop(0)
+            available_agents_len -= 1
+
+            # temp holder for used agents, so that the same connection wont be made twice
+            temp_agents_holder = []
+
+            # creating all of first's connections
+            for _ in range(remaining_contacts[current_agent]):
+                if connections_sum <= 1 or available_agents_len <= 1:
+                    break
+
+                # choosing 2nd agent for the connection
+                # todo make sure the % operator doesn't harm the randomness too much
+                roll_index = rolls[roll_cnt] % available_agents_len
+                roll_cnt += 1
+                second_agent = available_agents[roll_index]
+
+                # adding the newly made connection to the to-be-added connections structure
+                connections_tuples[0, connections_cnt] = current_agent.index
+                connections_tuples[1, connections_cnt] = second_agent.index
+                connections_cnt += 1
+                connections_tuples[0, connections_cnt] = second_agent.index
+                connections_tuples[1, connections_cnt] = current_agent.index
+                connections_cnt += 1
+
+                # note that there is no reason to update firs's remaining connections for efficiency
+                remaining_contacts[second_agent] -= 1
+                connections_sum -= 2
+
+                if remaining_contacts[second_agent] <= 0:
+                    del (available_agents[roll_index])
```

Comment:
deleting an item from the middle of an array is incredibly heavy, and requires O(N) operations

Reply:
we must keep track on who still lack connections.
the only other way i see is choosing a random agent from the dict at each iterate, which means activating random each loop, and deleting items straight from the dictionary.
do you think it will be better?